pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,343 @@
+
+	// Patch felix configuration if necessary.
+	err = r.patchFelixTproxyMode(ctx, applicationLayer)
+
+	if err != nil {
+		reqLogger.Error(err, "Error patching felix configuration")
+		r.status.SetDegraded("Error patching felix configuration", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	lcSpec := applicationLayer.Spec.LogCollection
+	component := applicationlayer.ApplicationLayer(pullSecrets, installation, rmeta.OSTypeLinux,
+		r.isLogsCollectionEnabled(lcSpec), lcSpec.LogIntervalSeconds, lcSpec.LogRequestsPerInterval)
+
+	ch := utils.NewComponentHandler(log, r.client, r.scheme, applicationLayer)
+
+	if err = imageset.ApplyImageSet(ctx, r.client, variant, component); err != nil {
+		reqLogger.Error(err, "Error with images from ImageSet")
+		r.status.SetDegraded("Error with images from ImageSet", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// TODO: when there are more ApplicationLayer options then it will need to be restructured, as each of the
+	// different features will not have their own CreateOrUpdateOrDelete
+	if err = ch.CreateOrUpdateOrDelete(ctx, component, r.status); err != nil {
+		reqLogger.Error(err, "Error creating / updating resource")
+		r.status.SetDegraded("Error creating / updating resource", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// Clear the degraded bit if we've reached this far.
+	r.status.ClearDegraded()
+
+	if !r.status.IsAvailable() {
+		// Schedule a kick to check again in the near future, hopefully by then things will be available.
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// Everything is available - update the CRD status.
+	applicationLayer.Status.State = operatorv1.TigeraStatusReady
+	if err = r.client.Status().Update(ctx, applicationLayer); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// updateApplicationLayerWithDefaults populates the applicationlayer with defaults.
+func updateApplicationLayerWithDefaults(al *operatorv1.ApplicationLayer) {
+	defaultLogIntervalSeconds := int64(5)
+	defaultLogRequestsPerInterval := int64(-1)
+
+	if al.Spec.LogCollection != nil {
+		if al.Spec.LogCollection.LogRequestsPerInterval == nil {
+			al.Spec.LogCollection.LogRequestsPerInterval = &defaultLogRequestsPerInterval
+		}
+		if al.Spec.LogCollection.LogIntervalSeconds == nil {
+			al.Spec.LogCollection.LogIntervalSeconds = &defaultLogIntervalSeconds
+		}
+	}
+}
+
+// validateApplicationLayer validates ApplicationLayer
+func validateApplicationLayer(al *operatorv1.ApplicationLayer) error {
+	lcSpec := al.Spec.LogCollection
+
+	// If ApplicationLayer spec exists then LogCollection should be set.
+	// TODO: when we will have multiple features in future this should change to at least one feature being set
+	if lcSpec == nil {
+		return fmt.Errorf("missing required LogCollection spec in applicationLayer")
+	}
+
+	return nil
+}
+
+// getApplicationLayer returns the default ApplicationLayer instance with defaults populated.
```

Comment:
This comment is not accurate, we don't handle defaulting in this function.

Reply:
fixed in 6c8d0e4